hypothesis-python/tests/cover/test_database_backend.py

```diff
@@ -500,3 +509,245 @@ def test_uleb_128_roundtrips(n1):
     idx, n2 = _unpack_uleb128(buffer1)
     assert idx == len(buffer1)
     assert n1 == n2
+
+
+def _database_conforms_to_listener_api(
+    create_db,
+    *,
+    flush=None,
+    supports_move=False,
+    supports_value_delete=True,
+    save_on_no_k1_in_move=False,
+    parent_settings=None,
+):
+    # this function is a big mess to support a bunch of different special cases
+    # for different databases, sorry. In return, we get one big stateful test
+    # we can use to test the listener api for all of our databases.
+    #
+    # * create_db is a callable which accepts one argument (a path to a temporary
+    #   directory) and returns a database instance.
+    # * flush is a callable which takes the instantiated db as an argument, and
+    #   is called on every step as an invariant. This lets the database do things
+    #   like, time.sleep to give time for events to fire.
+    # * supports_move is True if the db fires a "move" event,
+    #   and False if it fires a "delete" followed by a "save" instead.
+    # * suports_value_delete is True if the db supports passing
+    #   the exact value of a deleted key in "delete" events. The directory database
+    #   notably does not support this, and passes None instead.
+    # * save_on_no_k1_in_move is True if the db fires a "save" rather than "move" if
+    #   the value is not present in the source key.
```

Comment:
At this point I'm tempted to say that we _never_ emit `move` events, only `save`+`delete`.  It's a little bit more fiddly for database implementors but easier for listeners, and I like that tradeoff...

Reply:
yeah `move` has been a pain to implement (remaining failure is linux emitting move events differently than macos/windows under certain circumstances, I think), and the performance gains are minimal, so I'm fine with this 👍